### DISCUSSION TITLE
Keycloak doesnt work through gateway

### DISCUSSION THREAD

**Discussion by kowiste** [keycloak/keycloak#10476] [Q&A]
Hello,
I have keycloak inside a docker container and a microservice architecture, in which I have a gateway with multiple endpoints. In those endpoints I use a middleware in which I made a request to "/auth/realms/testrealm/protocol/openid-connect/userinfo" so if the token is valid this endpoint give me a 200 and the request follow to the services, if is not valid the gateway return a 401.

So if I ask for a token directly to the keycloak container in "/auth/realms/testrealm/protocol/openid-connect/token"  everything go fine and works.
Now I create an endpoint to ask for a token through the gateway, I have a response with the token but when I use it in an endpoint I receive :
<code>"{\"error\":\"invalid_token\",\"error_description\":\"Token verification failed\"}"</code>

The only different in the token that I can see is : 

JWT token receive directly from the container
<code>
 "iss": "http://localhost:8080/auth/realms/testrealm"
</code>

JWT token receive directly from the container
<code>
"iss": "http://localhost:5001/auth/realms/testrealm"
</code>

8080 is the keycloak service and 5001 gateway.

What can I do to make it work? I created the /auth/realms/testrealm endpoint just in case, but is not call.

**Comment by stianst** [comment]
Take a look at https://www.keycloak.org/server/reverseproxy
